custom_components/climate_template/climate.py

```diff
@@ -729,72 +764,81 @@ def swing_modes(self):
 
     async def async_set_hvac_mode(self, hvac_mode: str) -> None:
         """Set new operation mode."""
-        if self._hvac_mode_template is None:
-            self._current_operation = hvac_mode  # always optimistic
+        if hvac_mode != self._current_operation:  # Only proceed if there's a change
+            self._current_operation = hvac_mode
+            if self._set_hvac_mode_script is not None:
+                await self._set_hvac_mode_script.async_run(
+                    run_variables={ATTR_HVAC_MODE: hvac_mode}, context=self._context
+                )
             self.async_write_ha_state()
 
-        if self._set_hvac_mode_script is not None:
-            await self._set_hvac_mode_script.async_run(
-                run_variables={ATTR_HVAC_MODE: hvac_mode}, context=self._context
-            )
-
     async def async_set_preset_mode(self, preset_mode: str) -> None:
         """Set new preset mode."""
-        if self._preset_mode_template is None:
+        if preset_mode != self._current_preset_mode:  # Only proceed if there's a change
             self._current_preset_mode = preset_mode
+            if self._set_preset_mode_script is not None:
+                await self._set_preset_mode_script.async_run(
+                    run_variables={ATTR_PRESET_MODE: preset_mode}, context=self._context
+                )
             self.async_write_ha_state()
 
-        if self._set_preset_mode_script is not None:
-            await self._set_preset_mode_script.async_run(
-                run_variables={ATTR_PRESET_MODE: preset_mode}, context=self._context
-            )
-
     async def async_set_fan_mode(self, fan_mode: str) -> None:
         """Set new fan mode."""
-        if self._fan_mode_template is None:
-            self._current_fan_mode = fan_mode  # always optimistic
+        if fan_mode != self._current_fan_mode:  # Only proceed if there's a change
+            self._current_fan_mode = fan_mode
+            if self._set_fan_mode_script is not None:
+                await self._set_fan_mode_script.async_run(
+                    run_variables={ATTR_FAN_MODE: fan_mode}, context=self._context
+                )
             self.async_write_ha_state()
 
-        if self._set_fan_mode_script is not None:
-            await self._set_fan_mode_script.async_run(
-                run_variables={ATTR_FAN_MODE: fan_mode}, context=self._context
-            )
-
     async def async_set_swing_mode(self, swing_mode: str) -> None:
         """Set new swing mode."""
-        if self._swing_mode_template is None:  # use optimistic mode
+        if swing_mode != self._current_swing_mode:  # Only proceed if there's a change
             self._current_swing_mode = swing_mode
+            if self._set_swing_mode_script is not None:
+                await self._set_swing_mode_script.async_run(
+                    run_variables={ATTR_SWING_MODE: swing_mode}, context=self._context
+                )
```

Comment:
Is there a good reason to only run the script when the state changes in HA, I don't think there's really any harm in running it every time, especially if the script its running can sometimes fail to send an update.

This integration should be based on what home assistant does internally for template entities, the fan template is the closest and they too run the script and write the state on every action.
https://github.com/home-assistant/core/blob/dev/homeassistant/components/template/fan.py

Reply:
> Is there a good reason to only run the script when the state changes in HA, I don't think there's really any harm in running it every time, especially if the script its running can sometimes fail to send an update.
> 
> This integration should be based on what home assistant does internally for template entities, the fan template is the closest and they too run the script and write the state on every action. https://github.com/home-assistant/core/blob/dev/homeassistant/components/template/fan.py

Fixed this

Reply:
> Is there a good reason to only run the script when the state changes in HA, I don't think there's really any harm in running it every time, especially if the script its running can sometimes fail to send an update.
> 
> This integration should be based on what home assistant does internally for template entities, the fan template is the closest and they too run the script and write the state on every action. https://github.com/home-assistant/core/blob/dev/homeassistant/components/template/fan.py

@jcwillox Fixed this to bring it in line with HA and updated the PR.